### PR TITLE
Bratseth/deployment scheduling with blocking

### PR DESCRIPTION
**Pull request by bratseth**
@mpolden please review

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 1]
```diff
@@ -67,20 +72,18 @@ public void triggerFromCompletion(JobReport report) {
...
+                            application = application.withDeploying(Optional.of(Change.ApplicationChange.unknown()));
+                    }
+                    else { // postpone
                         applications().store(application.withOutstandingChange(true), lock);
                         return;
-                    } else { // start a new change deployment
-                        application = application.withDeploying(Optional.of(Change.ApplicationChange.unknown()));
                     }
                 } 
-                else if (order.isLastBeforeProduction(report.jobType()) && application.deployingBlocked(clock.instant())) {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: Can the `isLastBeforeProduction` method be removed then?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java [hunk 2]
```diff
@@ -10,6 +10,7 @@
 import com.yahoo.config.provision.Zone;
 import com.yahoo.vespa.hosted.controller.Controller;
 
+import java.time.Duration;
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java]: Unused import?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/ApplicationList.java [hunk 3]
```diff
@@ -12,6 +12,7 @@
 import java.time.Instant;
 import java.util.Comparator;
 import java.util.List;
+import java.util.Optional;
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/ApplicationList.java]: Unused import?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java [hunk 4]
```diff
@@ -14,7 +14,9 @@
 import com.yahoo.vespa.hosted.controller.application.DeploymentJobs;
 import com.yahoo.vespa.hosted.controller.application.DeploymentJobs.JobReport;
 import com.yahoo.vespa.hosted.controller.application.DeploymentJobs.JobType;
+import com.yahoo.vespa.hosted.controller.application.JobStatus;
 
+import java.time.Duration;
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java]: Unused imports?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/JobStatus.java [hunk 5]
```diff
@@ -98,6 +99,14 @@ else if (! lastTriggered.isPresent()) {
 
     /** Returns true unless this job last completed with a failure */
     public boolean isSuccess() { return ! jobError.isPresent(); }
+    
+    /** Returns true if last triggered is newer than last completed and was started after timeoutLimit */
+    public boolean isRunning(Instant timeoutLimit) {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/JobStatus.java]: Isn't this the same as the existing `inProgress`, the only difference being that this supports a timeout limit? Consider merging the methods or at least using the same name for both.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 6]
```diff
@@ -106,12 +109,64 @@ else if (shouldRetryNow(application))
     }
 
     /**
+     * Find jobs that can and should run but is currently not.
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: "is" -> "are"?

**Review by mpolden — CHANGES_REQUESTED** [review]

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/JobStatus.java (hunk 5 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/JobStatus.java]: Good catch. Yes we should merge them and always pass a timeoutLimit.
I see that at least some of the uses of inProgress solved that problem separately so there is an opportunity for simplification here, however, I also notice there are no tests of those particular alternatives, so I need to add that first. So, this becomes a substantial change in itself, I want to do it in a separate PR so I added a TODO for now.

The other things are addressed.

Comment (mpolden) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/JobStatus.java]: Great!

**Pull request merged**